<a href="https://colab.research.google.com/github/David-gutierrez-rodriguez/Simulacion-I/blob/main/Eficiencia_MC_crudo_con_variables_antit%C3%A9ticas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Se dice que el método 1 es más eficiente que el método 2 si
$$\varepsilon = \frac{var(\theta_1)t_1}{var(\theta_2)t_2} < 1$$

In [19]:
import numpy as np
from random import random
import matplotlib.pyplot as plt
import time

In [20]:
def g(x):
  return 1/(1 + x**2)

Método de Monte Carlo crudo

In [21]:
def MC_crudo(a, b, N):
  I = 0
  for i in range(N):
    I += g(random()*(b-a)+a)
  return I/N

Método de Monte Carlo con variables antitéticas

In [22]:
def MC_crudo_a(a, b, N):
  I1, I2 = 0, 0
  for i in range(N):
    r = random()
    I1 += g(r*(b-a)+a)
    I2 += g((1-r)*(b-a)+a)
  return (I1 + I2)/(2*N)

Tiempo de ejecución de un método

In [23]:
def tiempo(metodo, n, N):
  t_f = time.time()
  valores = [metodo(-1, 1, n) for _ in range(N)]
  t_i = time.time()
  return np.mean(valores), np.var(valores), -t_f + t_i

Realizamos el experimento con $n = 1000$ muestras y lo repetimos $N = 100$ veces en ambos casos

In [28]:
theta_1, var_theta_1, t_1 = tiempo(MC_crudo_a, 1000, 1000)
theta_2, var_theta_2, t_2 = tiempo(MC_crudo, 1000, 1000)

In [37]:
epsilon = (t_1*var_theta_1)/(t_2*var_theta_2)

print(f"theta_1 = {theta_1}")
print(f"var(theta_1) = {var_theta_1}")
print(f"t_1 = {t_1:.2f}\n")
print(f"theta_2 = {theta_2}")
print(f"var(theta_2) = {var_theta_2}")
print(f"t_2 = {t_2:.2f}\n")
print(f"epsilon = {epsilon:.2f}")
print(f"\nReducción de varianza: {100*abs(var_theta_1 - var_theta_2)/var_theta_1:.2f}%")

theta_1 = 0.7856140023950513
var(theta_1) = 2.5687223891245316e-05
t_1 = 0.63

theta_2 = 0.7856621799865434
var(theta_2) = 2.8571110486155504e-05
t_2 = 0.30

epsilon = 1.87

Reducción de varianza: 11.23%


Se concluye que el método de Monte Carlo crudo es más eficiente para esta simulación. Aunque la técnica de variables antitéticas redujo la varianza en un 11%, este beneficio fue superado por un tiempo de ejecución que se duplicó. El indicador de eficiencia relativa, $\varepsilon$ = 1.85, confirma que el costo computacional adicional no justifica la ganancia en precisión, haciendo del método estándar la opción superior.